[OpenMP] Add technical documentation (#4)
HEAPLab/TAFFO: doc/OpenMPSupport.md

# TAFFO OpenMP Support

> TAFFO has preliminary support for [OpenMP](https://www.openmp.org) parallel constructs.

Reviewer: ```suggestion
TAFFO has preliminary support for [OpenMP](https://www.openmp.org) parallel constructs.
We classify the support for most popular OpenMP features in **fully supported**, **partially supported**, and **not supported**.
```

Reviewer 2: After the latest _improvements_ (automatic disabling of conversion for shared variables), the **partially supported** (which had enormous problems) and the **not_supported** sets are now merged into one, therefore I would avoid doing further divisions. 

The main difference, which I will better explain in the user guide, is:
- supported parallel constructs have full support for taffo conversions of variables
- unsupported parallel constructs do not break anything, but no shared variable is converted in fixed-points

The goals of the integration are the following:

> - The usage of a simple workflow to easily patch the runtime OpenMP function during optimizations,

Reviewer: What do you mean by `runtime OpenMP function`?

Reviewer 2: Following [`OpenMP LLVM documentation`](https://openmp.llvm.org/Reference.pdf), The runtime OpenMP library should allow to interact with the OpenMP runtime with the related functions (like the `__kmpc_fork_call`). The library in particular is called "OpenMP Runtime library". I will insert an example to give a better idea 👌

>  requiring the least amount of work necessary for intermediary passes.

Reviewer: ```suggestion
 requiring the least amount of work necessary for intermediate passes.
```
?

- Basic support for the most common OpenMP directives, such as `omp parallel for`. 

>  This allows TAFFO to convert the floating-point logic inside the supported OpenMP pragmas to fixed-point.

Reviewer: ```suggestion
 This allows TAFFO to convert the floating-point code scoped by the supported OpenMP pragmas into fixed-point.
```

> - Conservativeness in handling not supported directives, such as `omp task`, to avoid failures.

Reviewer: ```suggestion
- Preserving functional behavior of the application with not supported directives, such as `omp task`, to avoid failures.
```

- Evaluate the results with benchmarks adapted to OpenMP and containing floating-points.

Note that a thorough analysis of the effects of data-dependency in the parallel execution is out of the scope of the integration.

> **If you are a TAFFO user and not interested in the technical details**, note the following:

Reviewer: I suggest to re-word the sentence as in "More technical details for developers are available elsewhere...."

Reviewer: Move to another document `elsewhere` implementation details.

Reviewer 2: Split the old document in UserGuide, DevGuide, and benchmarks

 In particular, the original function is saved in a local variable using a bitcast instruction. 
 This is necessary because otherwise the following optimization passes, such as Dead Code Elimination (DCE), 
 would remove the unused original function, making it impossible to reinsert the original calls in the Conversion pass.

#### Range Analysis and Data Type Allocation passes
TAFFO will regularly analyze the trampoline calls,  without any modification to these intermediate passes.

#### Conversion pass
At the end of the Conversion pass, it is necessary to transform back the trampolines into their original calls. 
To do so, the original function is picked from the saved bitcast and placed in place of the trampoline call, which will be deleted. 
The original function will have the relevant arguments replaced accordingly, specifically the indirect function to be called and the shared variables needed. 

In the above example, TAFFO will correctly convert the floating-point instructions in fixed-points, and
the call to `@__kmpc_fork_call` will contain `@.omp_outlined..2_fixp` instead of `@.omp_outlined.` and `i32* %foo.s6_26fixp` instead of `float* %foo`.

> ## Benchmarks

Reviewer: The benchmark evaluation should go into a new document.
It is more of a one-time thing, rather than a document we want to keep up to date

Reviewer 2: Moved in another document
